apiserver/facades/agent/storageprovisioner/storageprovisioner.go

```diff
@@ -1700,3 +1705,55 @@ func (s *StorageProvisionerAPIv4) RemoveAttachment(ctx context.Context, args par
 	}
 	return results, nil
 }
+
+// SetStatus sets the status of each given storage artefact.
+func (s *StorageProvisionerAPIv4) SetStatus(ctx context.Context, args params.SetStatus) (params.ErrorResults, error) {
+	result := params.ErrorResults{
+		Results: make([]params.ErrorResult, len(args.Entities)),
+	}
+	if len(args.Entities) == 0 {
+		return result, nil
+	}
+	canModify, err := s.getStorageEntityAuthFunc(ctx)
+	if err != nil {
+		return params.ErrorResults{}, err
+	}
+	now := s.clock.Now()
+	for i, arg := range args.Entities {
+		tag, err := names.ParseTag(arg.Tag)
+		if err != nil {
+			result.Results[i].Error = apiservererrors.ServerError(err)
+			continue
+		}
+		if !canModify(tag) {
+			result.Results[i].Error = apiservererrors.ServerError(apiservererrors.ErrPerm)
+			continue
+		}
+		sInfo := status.StatusInfo{
+			Status:  status.Status(arg.Status),
+			Message: arg.Info,
+			Data:    arg.Data,
+			Since:   &now,
+		}
+		var statusErr error
+		switch tag := tag.(type) {
+		case names.FilesystemTag:
+			filesystemId := idFromTag(tag.Id())
+			statusErr = s.storageStatusService.SetVolumeStatus(ctx, filesystemId, sInfo)
+		case names.VolumeTag:
+			volId := idFromTag(tag.Id())
+			statusErr = s.storageStatusService.SetVolumeStatus(ctx, volId, sInfo)
+		default:
+			statusErr = apiservererrors.ErrPerm
+		}
+		result.Results[i].Error = apiservererrors.ServerError(statusErr)
+	}
+	return result, nil
+}
+
+func idFromTag(tag string) string {
+	// filesystem or volume tags always end with "/N"
+	// where N is the filesystem or volume id that we want.
+	pos := strings.LastIndex(tag, "/")
+	return tag[pos:]
+}
```

Comment:
For storage imported via model migration, we need to maintain consistent filesystem and volume IDs. We are not doing that currently.

I suggest we drop this. Just pass the id along to the service, this will then highlight where we have messed up migration logic.